sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PrivateFieldUsedAsLocalVariable.cs

```diff
@@ -36,340 +36,216 @@ namespace SonarAnalyzer.Rules.CSharp
-            ISet<ISymbol> classMethods)
-        {
-            var referencedAtLeastOnceFromClassMethod = new HashSet<ISymbol>();
-
-            foreach (var references in referencesByEnclosingSymbol)
+            public bool IsRemovableField(ISymbol fieldSymbol)
             {
-                if (!classMethods.Contains(references.Key))
-                {
-                    foreach (var reference in references.Value)
-                    {
-                        privateFields[reference.Value].Excluded = true;
-                    }
+                var writesByMethod = writes.GetValueOrDefault(fieldSymbol);
+                var readsByMethod = reads.GetValueOrDefault(fieldSymbol);
 
-                    continue;
+                // No methods overwrite the field value
+                if (writesByMethod == null)
+                {
+                    return false;
                 }
 
-                foreach (var reference in references.Value)
-                {
-                    referencedAtLeastOnceFromClassMethod.Add(reference.Value);
+                // A field is removable when no method read it, or all methods that read it, first overwrite it
+                return readsByMethod == null
+                    || readsByMethod.Keys.All(MethodOverwritesBeforeReading);
 
-                    if (!IsReferenceToSingleFieldValue(reference))
-                    {
-                        privateFields[reference.Value].Excluded = true;
-                    }
+                bool MethodOverwritesBeforeReading(ISymbol methodSymbol)
+                {
+                    var writeStatements = writesByMethod.GetValueOrDefault(methodSymbol);
+                    var readStatements = readsByMethod.GetValueOrDefault(methodSymbol);
+
+                    // Note that Enumerable.All() will return true if readStatements is empty. The collection
+                    // will be empty if the field is read only in property/field initializers or returned from
+                    // expression-bodied methods.
+                    return readStatements == null
+                        || writeStatements != null && readStatements.All(IsPrecededWithWrite);
+
+                    bool IsPrecededWithWrite(StatementSyntax readStatement) =>
+                        GetPreviousStatements(readStatement)
+                            .Any(writeStatements.Contains);
                 }
-            }
 
-            foreach (var privateField in privateFields.Values)
-            {
-                if (!referencedAtLeastOnceFromClassMethod.Contains(privateField.Symbol))
+                // Returns all statements before the specified statement within the containing method
+                IEnumerable<StatementSyntax> GetPreviousStatements(StatementSyntax statement)
                 {
-                    privateField.Excluded = true;
+                    var previousStatements = statement.Parent.ChildNodes()
+                        .OfType<StatementSyntax>()

```

Comment:
I might say something a bit stupid but I would put the `TakeWhile` before the `OfType` because we will do less useless casts.